pkg/k8s/service_cache.go

```diff
@@ -97,12 +98,32 @@ type ServiceEvent struct {
 	SWG *lock.StoppableWaitGroup
 }
 
+// ServiceNotification is a slimmed down version of a ServiceEvent. In particular
+// notifications are optional and thus do not contain a wait group to allow
+// producers to wait for the notification to be consumed.
+type ServiceNotification struct {
+	Action       CacheAction
+	ID           ServiceID
+	Service      *Service
+	OldService   *Service
+	Endpoints    *Endpoints
+	OldEndpoints *Endpoints
+}
+
 // ServiceCache is a list of services correlated with the matching endpoints.
 // The Events member will receive events as services.
 type ServiceCache struct {
 	config ServiceCacheConfig
 
-	Events chan ServiceEvent
+	// Events may only be read by single consumer. The consumer must acknowledge
+	// every event by calling Done() on the ServiceEvent.SWG.
+	Events <-chan ServiceEvent
+
+	// notifications are multicast and may be received by multiple subscribers.
+	notifications stream.Observable[ServiceNotification]
+
+	emitEvent func(ServiceEvent)
```

Comment:
Is there a reason this is overrideable?

Reply:
Because it's a value returned by `stream.Multicast[T]`

Reply:
`emitNotification` is, but not `emitEvent`. This is extremely minor splitting hairs, but I would expect `emitEvent()` to be a method, not a function pointer.

Reply:
Ah, apologies, the code changes a few times, it used to be `emitNotifications` at some point.

 For `emitEvent` the reason for being a function pointer is that was a convenient way to not have to introduce a separate `chan<- ServiceEvent` field alongside `emitNotification`. But that's easily changed, as I don't feel strongly about this.